File: examples/ecommerce-ts/handlers/get_product.ts
import type { RouteRequest, RouteResult, PrefetchMessage, PrefetchDb, HandleContext, RenderContext } from "tiger-web";
import { esc, price } from "tiger-web";

// [route] .get_product
// match GET /products/:id
export function route(req: RouteRequest): RouteResult | null {
  return { operation: "get_product", id: req.params.id };
}

// [prefetch] .get_product
export function prefetch(msg: PrefetchMessage, db: PrefetchDb) {
  const product = db.query("SELECT id, name, description, price_cents, inventory, version, active FROM products WHERE id = ?1", msg.id);
  return { product };
}

// [handle] .get_product
export function handle(ctx: HandleContext): string {
  if (!ctx.prefetched.product) return "not_found";
  if (!ctx.prefetched.product.active) return "not_found";
  return "ok";
}

// [render] .get_product
export function render(ctx: RenderContext): string {
  switch (ctx.status) {
    case "ok": {
      const p = ctx.prefetched.product;
      return `<div class="card"><strong>${esc(p.name)}</strong> &mdash; ${price(p.price_cents)}` +
        ` &mdash; inv: ${p.inventory} &mdash; v${p.version}` +
        (!p.active ? ` <span class="error">[inactive]</span>` : ``) +
        `<div class="meta">${p.id}</div>` +
        (p.description ? `<div class="meta">${esc(p.description)}</div>` : ``) +
        `</div>`;
    }
    case "not_found":
      return '<div class="error">Product not found</div>';
  }
  throw new Error("unreachable: " + ctx.status);
}

File: examples/ecommerce-py/handlers/get_product.py
import sys
sys.path.insert(0, str(__import__("pathlib").Path(__file__).resolve().parents[2] / "adapters" / "python"))
from tiger_web import esc, price


def route(req):
    return {"operation": "get_product", "id": req["params"]["id"]}


def prefetch(msg, db):
    product = db.query(
        "SELECT id, name, description, price_cents, inventory, version, active "
        "FROM products WHERE id = ?1", msg["id"]
    )
    return {"product": product}


def handle(ctx, db):
    if not ctx["prefetched"].get("product"):
        return "not_found"
    if not ctx["prefetched"]["product"].get("active"):
        return "not_found"
    return "ok"


def render(ctx, db=None):
    if ctx["status"] == "not_found":
        return '<div class="error">Product not found</div>'

    p = ctx["prefetched"]["product"]
    desc = f'<div class="meta">{esc(p["description"])}</div>' if p.get("description") else ""
    active = "" if p.get("active") else ' <span class="error">[inactive]</span>'

    return (
        f'<div class="card"><strong>{esc(p["name"])}</strong> &mdash; {price(p["price_cents"])}'
        f' &mdash; inv: {p["inventory"]} &mdash; v{p["version"]}{active}'
        f'<div class="meta">{p["id"]}</div>{desc}</div>'
    )
